onnxruntime/core/framework/execution_provider.cc

```diff
@@ -19,6 +19,7 @@ inline int MakeKey(int id, OrtMemType mem_type) {
 }  // namespace
 
 AllocatorPtr IExecutionProvider::GetAllocator(int device_id, OrtMemType mem_type) const {
+  ORT_ENFORCE(device_id == 0 || mem_type == OrtMemType::OrtMemTypeDefault, "It's illegal for non-CPU device to access CPU-related memory type");
```

Comment:
It seems not the check we want to do. I think the case we want to check is whether there is a call that search on OrtMemTypeDefault but the device_id doesn't match ExecutionProvider::GetDeviceId(). The main purpose is to check whether we have any place that the developer just hardcode the device_id to 0 on default memory type.